### PR TITLE
Add _scaled_dot_product_attention_math_native_impl op

### PR DESCRIPTION
**Pull request by AlienLiang23**
Add _scaled_dot_product_attention_math_native_impl op

skip TestCommonXPU::test_dtypes_nn_functional_scaled_dot_product_attention_xpu

explicitly register _softmax_out op & _softmax_backward_data_out op & tril_out op

### CODE REVIEW

File: src/aten/SoftMax.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#include <ATen/ATen.h>
+#include <ATen/XPUNativeFunctions.h>
+#include <ATen/core/Tensor.h>
+
+namespace at {
+
+// Following three functions are currently not use by sdpa
```
Comment (fengyuan14) [inline src/aten/SoftMax.cpp:7]: Please check, if stock CUDA registers, we register.

Comment (fengyuan14) [reply src/aten/SoftMax.cpp:7]: Even our highest priority is HF/TIMM/TB, we are gradually providing on-par op coverage as CUDA. If you are implementing an operator, you need register and implement its variants as well. It won't take your a lot efforts.

File: src/aten/transformers/attention.cpp
Comment (fengyuan14) [inline src/aten/transformers/attention.cpp:1]: Please aligh stock CPU and stock CUDA sdp math registration. Remove it, if it is unnecessary.

File: src/aten/SoftMax.cpp [hunk 2]
```diff
@@ -1,40 +1,84 @@
...
     Tensor& out) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device, out, "xpu::_softmax_out", "out");
+  c10::impl::check_and_update_common_device(
+      common_device, self, "xpu::_softmax_out", "self");
+  torch_meta_func__softmax(out, self, dim, half_to_float);
   Tensor out_cpu = out.to(Device(kCPU));
   Tensor self_cpu = self.to(Device(kCPU));
+  // std::cout << self_cpu.data() << std::endl;
```
Comment (fengyuan14) [inline src/aten/SoftMax.cpp:46]: Remove unnecessary codes.

File: src/aten/TriangularOps.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#include <ATen/ATen.h>
+#include <ATen/XPUNativeFunctions.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/core/op_registration/adaption.h>
+// #include <c10/core/SymIntArrayRef.h>
```
Comment (fengyuan14) [inline src/aten/TriangularOps.cpp:5]: Remove the line, if unnecessary.

File: src/aten/TriangularOps.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#include <comm/RegisterUtils.h>
+// #include <torch/library.h>
```
Comment (fengyuan14) [inline src/aten/TriangularOps.cpp:7]: Ditto

File: src/aten/TriangularOps.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace at {
+
+void torch_meta_func_tril(const Tensor& self, int64_t k, Tensor& out) {
```
Comment (fengyuan14) [inline src/aten/TriangularOps.cpp:11]: Ditto. Naming.

File: src/aten/SoftMax.cpp [hunk 6]
```diff
@@ -1,40 +1,84 @@
...
   at::_softmax_out(out_cpu, self_cpu, dim, half_to_float);
+  // std::cout << out_cpu.data() << std::endl;
```
Comment (fengyuan14) [inline src/aten/SoftMax.cpp:48]: Ditto

File: src/aten/SoftMax.cpp [hunk 7]
```diff
@@ -1,40 +1,84 @@
...
 namespace at {
 
-Tensor XPUNativeFunctions::_softmax(
-    const Tensor& self,
-    int64_t dim,
-    bool half_to_float) {
-  Tensor self_cpu = self.to(Device(kCPU));
-  Tensor out = at::_softmax(self_cpu, dim, half_to_float);
-  return out.to(Device(kXPU));
+void torch_meta_func__softmax(
```
Comment (fengyuan14) [inline src/aten/SoftMax.cpp:11]: Align naming. No `torch_` prefix.

File: yaml/xpu_functions.yaml [hunk 8]
```diff
@@ -202,12 +202,8 @@ supported:
   - any.all_out
   - argmax.out
   - _local_scalar_dense
-  - _softmax
   - _softmax.out
-  - _softmax_backward_data
   - _softmax_backward_data.out
-  - tril
```
Comment (fengyuan14) [inline yaml/xpu_functions.yaml]: Why removing the variants?

**Review by fengyuan14 — APPROVED** [review]

**Comment by AlienLiang23** [comment]
image
TestCommonXPU::test_dtypes_nn_functional_scaled_dot_product_attention_xpu is skipped in pre-ci due to oneDNN matmul

File: test/xpu/run_test.py [hunk 9]
```diff
@@ -27,707 +27,765 @@ def launch_test(test_case, skip_list=None, exe_list=None):
...
-    "test_python_ref_executor__refs_special_multigammaln_mvlgamma_p_3_executor_aten_xpu_float64",
-    "test_numpy_ref_nn_functional_rms_norm_xpu_complex128",
-    "test_python_ref__refs_square_xpu_complex128",
-    "test_python_ref__refs_square_xpu_complex64",
-    "test_python_ref_executor__refs_istft_executor_aten_xpu_complex128",
-    "test_python_ref_executor__refs_square_executor_aten_xpu_complex128",
-    "test_python_ref_torch_fallback__refs_square_xpu_complex128",
-    "test_python_ref_torch_fallback__refs_square_xpu_complex64",
+# # test_ops
+# skip_list = (
```
Comment (fengyuan14) [inline test/xpu/run_test.py:31]: Please retrieve all cases.

File: test/xpu/run_test.py [hunk 10]
```diff
@@ -729,5 +729,65 @@ def launch_test(test_case, skip_list=None, exe_list=None):
 )
 res += launch_test("test_foreach_xpu.py", exe_list=execute_list)
 
+
+# test_transformers
+skip_list = (
+    # AssertionError: False is not true
+    # due to transformer_encoder_layer_forward op issue
+    "test_disable_fastpath_xpu",
+    # RuntimeError not raised by <lambda>
```
Comment (fengyuan14) [inline test/xpu/run_test.py:738]:
```suggestion
    # We have no mechanism to handle SDPBackend::ERROR so far.
```

File: test/xpu/run_test.py [hunk 11]
```diff
@@ -729,5 +729,65 @@ def launch_test(test_case, skip_list=None, exe_list=None):
...
+    "test_dispatch_fails_no_backend_xpu",
+    # Could not run 'aten::_to_copy' with arguments from the 'NestedTensorXPU' backend
+    # We have no mechanism to handle SDPBackend::ERROR so far.
```
Comment (fengyuan14) [inline test/xpu/run_test.py:741]:
```suggestion
```

File: test/xpu/run_test.py [hunk 12]
```diff
@@ -729,5 +729,65 @@ def launch_test(test_case, skip_list=None, exe_list=None):
 )
 res += launch_test("test_foreach_xpu.py", exe_list=execute_list)
 
+
+# test_transformers
+skip_list = (
+    # AssertionError: False is not true
+    # due to transformer_encoder_layer_forward op issue
```
Comment (fengyuan14) [inline test/xpu/run_test.py:736]:
```suggestion
    # CPU fallback failure. To support aten::transformer_encoder_layer_forward with proper priority.
```

**Review by fengyuan14 — APPROVED** [review]

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**